### PR TITLE
[WIP] Unit Testing: Initial framework

### PR DESCRIPTION
**Pull request by MohamedAshiqrh**
Signed-off-by: Mohamed Ashiq Liyazudeen <mliyazud@redhat.com>

---
This change is [Reviewable](https://reviewable.io/reviews/gluster/gluster-kubernetes/184)

### CODE REVIEW

**Comment by MohamedAshiqrh** [comment]
@jarrpa @obnoxxx @humblec @raghavendra-talur @ramkrsna Please Review.

**Comment by MohamedAshiqrh** [comment]
@raghavendra-talur Nice points. Worked like charm. Thanks.

**Comment by MohamedAshiqrh** [comment]
Now it Just tests for yaml syntax err. Mock func to be added and gk-deploy-unit should be used for testing.

**Comment by MohamedAshiqrh** [comment]
Still 100 more lines to go. Have to add more validations and exit conditions. validation for example, We don't validate the Template file locations specified or the defaults. 
@jarrpa Should I add the changes to gk-deploy along with this PR or send a new PR?

**Comment by jarrpa** [comment]
@MohamedAshiqrh That should be a separate PR. :)

File: deploy/tests/test.sh [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  file="${1}"
+  TEST=$(python -c 'import yaml,sys;yaml.safe_load(sys.stdin)' < ${file} 2>/dev/null)
+  if [[ $? -ne 0 ]]; then
+    echo "Failed ${file}"
+  fi
+}
+
+# copies the source and adds the mock source. This will be used for testing
+cp ../gk-deploy gk-deploy-unit
+sed -i '16s/^/\nsource mock.sh/' gk-deploy-unit
```
Comment (obnoxxx) [inline deploy/tests/test.sh]: This works, but I don't like this copying and modifying of the script very much.

It should be sufficient to source the mock.sh in *THIS* script, right?
If I am mistaken, could we possibly use a checked-in wrapper script?

File: deploy/tests/test.sh [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+# call all the yaml files and check if they exist. Also call yaml syntax checking
+for i in $KUBE_GLUSTER_DEAMONSET $KUBE_HEKETI_DEPLOYMENT $KUBE_HEKETI_SA $OCP_GLUSTER_TEMPLATE $OCP_HEKETI_TEMPLATE $OCP_HEKETI_SA 
+do
+  if test "$(ls ${i})"
+  then
+	echo "$i is present"
+	# Add here if the new file has more than one yaml context
+        if [[ "${i}" != "${KUBE_HEKETI_DEPLOYMENT}" ]]; then
+	  check "${i}"
```
Comment (obnoxxx) [inline deploy/tests/test.sh]: indentation

File: deploy/tests/test.sh [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+KUBE_HEKETI_SA=../kube-templates/heketi-service-account.yaml 
+OCP_GLUSTER_TEMPLATE=../ocp-templates/glusterfs-template.yaml
+OCP_HEKETI_TEMPLATE=../ocp-templates/heketi-template.yaml
+OCP_HEKETI_SA=../ocp-templates/heketi-service-account.yaml
+TOPOLOGY=../topology.json.sample
+
+# syntax checking can not be done if file has more than one yaml context 
+# so split and call check
+# check kube-templates/heketi-deployment.yaml to understand what I am talking about
+split_yaml_check() {
```
Comment (obnoxxx) [inline deploy/tests/test.sh]: It should be possible to write this function so that the case of a single context (even without opening "---\n") would be a valid corner case. So the caller should not need to know and just call this function, ideally under a different name, like yaml_syntax_check(...) or so...

File: deploy/tests/test.sh (hunk 1 above)
Comment (obnoxxx) [inline deploy/tests/test.sh]: I tested it, see https://github.com/obnoxxx/gluster-kubernetes/tree/pr-184-plus . (This also contains a few cleanups in the formatting etc).

But there seem to be issues. the sourcing in the test script does not work immediately.

I noticed a problem: When you just comment out the creation of the patched script, the test runs a little longer but succeeds nonetheless even though no kubectl or oc commands are available.. :-o This needs to be fixed.

**Comment by obnoxxx** [comment]
This was superseded by #194 which was based on this ideas.
